File: examples/custom/font-awesome/fontawesome_icon.py
from bokeh.core.properties import Enum, Float, Bool
from bokeh.models.widgets.icons import AbstractIcon
from named_icon import NamedIcon

class FontAwesomeIcon(AbstractIcon):
    """ A "stock" icon based on FontAwesome. """

    __implementation__ = "fontawesome_icon.coffee"
    __dependencies__ = {"font-awesome": "^4.6.3"}

    icon_name = Enum(NamedIcon, default="check", help="""
    What icon to use. See http://fortawesome.github.io/Font-Awesome/icons/
    for the list of available icons.
    """)

    size = Float(None, help="""
    The size multiplier (1x, 2x, ..., 5x).
    """)

    flip = Enum("horizontal", "vertical", default=None, help="""
    Optionally flip the icon horizontally or vertically.
    """)

    spin = Bool(False, help="""
    Indicates a spinning (animated) icon. This value is ignored for
    icons that do not support spinning.
    """)

File: examples/custom/font-awesome/fontawesome_icon.ts
import {AbstractIcon, AbstractIconView} from "models/widgets/abstract_icon"
import * as p from "core/properties"

import "./fontawesome.less"

export class FontAwesomeIconView extends AbstractIconView {
  model: FontAwesomeIcon

  connect_signals(): void {
    super.connect_signals()
    this.connect(this.model.change, () => this.render())
  }

  render(): void {
    super.render()

    this.el.style.display = "inline"
    this.el.style.verticalAlign = "middle"
    this.el.style.fontSize = `${this.model.size}em`

    this.el.classList.add("bk-u-fa")
    this.el.classList.add(`bk-u-fa-${this.model.icon_name}`)

    if (this.model.flip != null)
      this.el.classList.add(`bk-u-fa-flip-${this.model.flip}`)

    if (this.model.spin)
      this.el.classList.add("bk-u-fa-spin")
  }
}

export namespace FontAwesomeIcon {
  export type Attrs = p.AttrsOf<Props>

  export type Props = AbstractIcon.Props & {
    icon_name: p.Property<string>
    size: p.Property<number>
    flip: p.Property<"horizontal" | "vertical" | null>
    spin: p.Property<boolean>
  }
}

export interface FontAwesomeIcon extends FontAwesomeIcon.Attrs {}

export class FontAwesomeIcon extends AbstractIcon {
  properties: FontAwesomeIcon.Props

  constructor(attrs?: Partial<FontAwesomeIcon.Attrs>) {
    super(attrs)
  }

  static initClass(): void {
    this.prototype.default_view = FontAwesomeIconView

    this.define<FontAwesomeIcon.Props>({
      icon_name: [ p.String,  "check" ], // TODO (bev) enum?
      size:      [ p.Number,  1       ],
      flip:      [ p.Any              ], // TODO (bev)
      spin:      [ p.Boolean, false   ],
    })
  }
}
FontAwesomeIcon.initClass()
